handlers/handlers.go

```diff
@@ -34,28 +36,71 @@ func (h *Handler) MemesIndex(w http.ResponseWriter, r *http.Request) {
 }
 
 // MemesCreate creates a meme
+// TODO: function too long
 func (h *Handler) MemesCreate(w http.ResponseWriter, r *http.Request) {
-	meme := &models.Meme{URL: r.FormValue("url")}
+	meme := &models.Meme{OriginalUrl: r.FormValue("url")}
+	resp, err := http.Get(meme.OriginalUrl)
+	if err != nil {
+		log.Fatalf("%s %s", "http error", meme.OriginalUrl)
+		return
+	}
+	if resp.StatusCode != 200 || resp.ContentLength > maxImageContentLength {
+		log.Fatalf("%s %s", "error status or response too big", resp.StatusCode)
+		return
+	}
+	contentTypeValid := false
+	for _, contentType := range allowedContentTypes {
+		if contentType == resp.Header.Get("Content-Type") {
+			contentTypeValid = true
+			break
+		}
+	}
+	if !contentTypeValid {
+		log.Fatalf("invalid content type header")
+		return
+	}
+	if err := meme.SetData(resp.Body, resp.Header.Get("Content-Type")); err != nil {
+		log.Fatal(err)
+	}
 
 	if err := h.db.Update(meme.Save); err != nil {
 		log.Fatal(err)
 	}
-
 	json.NewEncoder(w).Encode(meme)
 }
 
+// MemeGet serves a meme image
+// TODO: return appropriate Content-Type header - may be required?
+func (h *Handler) MemeGet(w http.ResponseWriter, r *http.Request) {
+	meme := h.findMeme(r)
+
+	h.db.View(meme.GetImage)                                             // fetch image
+	w.Header().Set("Content-Length", strconv.Itoa(meme.ContentLength())) // set length
+	w.Header().Set("Content-Type", meme.ContentType)                     // set mime
+	if err := meme.WriteImage(w); err != nil {
+		log.Fatal("error writing binary data")
+	}
+}
+
 // MemeDestroy destroys a meme
 func (h *Handler) MemeDestroy(w http.ResponseWriter, r *http.Request) {
-	memeID, err := strconv.ParseUint(mux.Vars(r)["memeID"], 10, 64)
+	meme := h.findMeme(r)
 
-	if err != nil {
+	if err := h.db.Update(meme.Destroy); err != nil {
 		log.Fatal(err)
 	}
+}
 
-	meme := &models.Meme{ID: uint64(memeID)}
-	if err := h.db.Update(meme.Destroy); err != nil {
+func (h *Handler) findMeme(r *http.Request) *models.Meme {
+	memeID, err := strconv.ParseUint(mux.Vars(r)["memeID"], 10, 64)
+	if err != nil {
 		log.Fatal(err)
+		return nil
```

Comment:
Technically `log.Fatal` stops the program so `return` isn't necessary but probably this will be useful in the future